BABEL-745 Support STDEV(), STDEVP(), VAR(), VARP() statistical T-SQL

pratikzode opened the pull request:
STDEV()
STDEV function is used in SQL to calculate the standard deviation of a set of values in a given column of a table. STDEV stands for "standard deviation," which is a statistical measure that represents the degree of variation of a set of values from its mean or average

STDEVP()
STDEVP function is similar to the STDEV function in SQL, but it calculates the standard deviation of an entire population rather than a sample of the population. "STDEVP" stands for "standard deviation of a population".

VAR()
VAR function is used in SQL to calculate the variance of a set of values in a given column of a table. "VAR" stands for "variance," which is a statistical measure that represents how much a set of values varies from its mean or average.

VARP()
VARP function is used in SQL to calculate the variance of an entire population rather than a sample of the population. "VARP" stands for "variance of a population"

Task: BABEL-745

pratikzode committed:
BABEL-745 Support STDEV(), STDEVP(), VAR(), VARP() statistical T-SQL aggregates

STDEV()
STDEV function is used in SQL to calculate the standard deviation of a set of values in a given column of a table. STDEV stands for "standard deviation," which is a statistical measure that represents the degree of variation of a set of values from its mean or average

STDEVP()
STDEVP function is similar to the STDEV function in SQL, but it calculates the standard deviation of an entire population rather than a sample of the population. "STDEVP" stands for "standard deviation of a population".

VAR()
VAR function is used in SQL to calculate the variance of a set of values in a given column of a table. "VAR" stands for "variance," which is a statistical measure that represents how much a set of values varies from its mean or average.

VARP()
VARP function is used in SQL to calculate the variance of an entire population rather than a sample of the population. "VARP" stands for "variance of a population"

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-745 Added Babel to upgrade

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-745 Added to upgrade 3.1->3.2

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-745 Modify test cases

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-745 Modify test cases

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on test/JDBC/expected/BABEL-745-vu-prepare.out:
```diff
@@ -0,0 +1,67 @@
+-- Create SalesData table
+CREATE TABLE SalesData (
+  Product varchar(50),
+  SalesAmount decimal(10,2)
```
your function signature is receiving float value. 
So please also test with float value

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql:
```diff
@@ -78,3 +78,38 @@ DROP FUNCTION sys.pg_extension_config_remove(REGCLASS);
 -- Reset search_path to not affect any subsequent scripts
 SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
 
+CREATE AGGREGATE STDEV(float8) (
```
please specify sys schema for your function definitions

xhfanhe replied:
And please put everything before line 73
"-- Drops the temporary procedure used by the upgrade script."

macwoody commented on test/JDBC/input/BABEL-745-vu-prepare.sql:
```diff
@@ -0,0 +1,65 @@
+  SELECT VAR(SalesAmount) AS SalesAmountStdev
+  FROM SalesData
+  WHERE Product = @Product
+  GROUP BY Product;
```
for all select statement you are using, please add order by clause

macwoody commented on test/JDBC/input/BABEL-745-vu-verify.sql:
```diff
@@ -0,0 +1,35 @@
+
+SELECT Product, VARP(CAST(SalesAmount AS float8)) AS SalesAmountStdev
+FROM SalesData
+GROUP BY Product
```
ORDER BY

macwoody commented on test/JDBC/input/BABEL-745-vu-prepare.sql:
```diff
@@ -0,0 +1,65 @@
+  SalesAmount decimal(10,2)
+);
+-- Insert some sample data
+INSERT INTO SalesData (Product, SalesAmount)
```
please bring PG's existing tests here in addition to the scenario you created

pratikzode committed:
BABEL-745 Modify test cases

Specify sys schema for your function definitions.
Add order by clause to test cases

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

surendravishnoi commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -407,6 +407,7 @@ openquery_upgrd
 BABEL-3657
 test_windows_alter_user
 BABEL-733
+BABEL-745
```
adding to schedule will not exercise 3.1 to 3.2, add it to 14.6 or 14.7 as well

pratikzode committed:
BABEL-745 Adding upgrade to 14.6 or 14.7

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-745 Deleting upgrade in 14.6

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on test/JDBC/expected/BABEL-745-vu-prepare.out:
```diff
@@ -0,0 +1,115 @@
+
+
+-- Create a view for SalesData
+CREATE VIEW v_SalesData AS
```
The view definition has to contain the function definition, just like below store procedures

macwoody commented on test/JDBC/expected/BABEL-745-vu-verify.out:
```diff
@@ -0,0 +1,116 @@
+~~END~~
+
+
+SELECT
```
What is intention of this test ?

lejaokri commented on test/JDBC/expected/BABEL-745-vu-verify.out:
```diff
@@ -0,0 +1,116 @@
+GO
+~~START~~
+float
+1666.6666666666667
```
quick question: these output were verified against sql server, right?

pratikzode replied:
Yes

pratikzode committed:
BABEL-745 Modify test cases

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on test/JDBC/expected/BABEL-745-vu-prepare.out:
```diff
@@ -0,0 +1,69 @@
+
+-- Create a view for SalesData
+CREATE VIEW v_SalesData AS
+SELECT Product, STDEV(SalesAmount) AS SalesAmount
```
why just stdev ? plz have all other functions covered by view as well.

pratikzode replied:
Yes Covered it in new commit

pratikzode committed:
BABEL-745 Modify test cases

Created views for STDEVP, VAR, VARP

Task: BABEL-745
Signed-off-by: pratikzode <pratik.zode@gmail.com>
